### PR TITLE
bpf-loader: add btfhub support

### PR DESCRIPTION
**Pull request by ocfox**
#150 alternative, better way.

### CODE REVIEW

File: bpf-loader/include/eunomia/eunomia-bpf.hpp [hunk 1]
```diff
@@ -132,11 +133,11 @@ class bpf_skeleton
...
-    [[nodiscard]] int open_from_json_config(
-        const std::string &json_str,
-        std::vector<char> bpf_object_buffer) noexcept;
-    [[nodiscard]] int open_from_json_config(
-        const std::string &json_package) noexcept;
+    [[nodiscard]] int open_from_json_config(const std::string &json_str,
+                                            std::vector<char> bpf_object_buffer,
+                                            char *btf_archive_path) noexcept;
+    [[nodiscard]] int open_from_json_config(const std::string &json_package,
+                                            char *btf_archive_path) noexcept;
```
Comment (yunwei37) [inline bpf-loader/include/eunomia/eunomia-bpf.hpp:140]: char *btf_archive_path = nullptr

File: bpf-loader/src/btf_helpers.c [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+        goto out;
+
+    strncpy(info->kernel_release, u.release, FIELD_LEN);
+    strncpy(info->arch, u.machine, FIELD_LEN);
+
+    while ((read = getline(&line, &len, f)) != -1) {
+        if (sscanf(line, ID_FMT, info->id) == 1)
+            continue;
+
+        if (sscanf(line, VERSION_FMT, info->version) == 1) {
```
Comment (yunwei37) [inline bpf-loader/src/btf_helpers.c:57]: maybe you should use snscanf

**Review by yunwei37 — CHANGES_REQUESTED** [review]

File: bpf-loader/src/eunomia_meta.cpp [hunk 3]
```diff
@@ -183,7 +188,8 @@ bpf_skeleton::open_from_json_config(
 }
 
 int
-bpf_skeleton::open_from_json_config(const std::string &json_package) noexcept
+bpf_skeleton::open_from_json_config(const std::string &json_package,
+                                    char *btf_archive_path) noexcept
```
Comment (ocfox) [inline bpf-loader/src/eunomia_meta.cpp:192]:
```suggestion
bpf_skeleton::open_from_json_config(const std::string &json_package,
                                    char *btf_archive_path = nullptr) noexcept
```

File: bpf-loader/src/eunomia_meta.cpp [hunk 4]
```diff
@@ -167,12 +168,16 @@ eunomia_object_meta::from_json_str(const std::string &j_str)
 }
 
 int
-bpf_skeleton::open_from_json_config(
-    const std::string &json_str, std::vector<char> bpf_object_buffer) noexcept
+bpf_skeleton::open_from_json_config(const std::string &json_str,
+                                    std::vector<char> bpf_object_buffer,
+                                    char *btf_archive_path) noexcept
```
Comment (ocfox) [inline bpf-loader/src/eunomia_meta.cpp:173]:
```suggestion
bpf_skeleton::open_from_json_config(const std::string &json_str,
                                    std::vector<char> bpf_object_buffer,
                                    char *btf_archive_path = nullptr) noexcept
```

File: bpf-loader/src/main.cpp [hunk 5]
```diff
@@ -26,7 +26,7 @@ main(int argc, char *argv[])
         (std::istreambuf_iterator<char>(bpf_file)),
         std::istreambuf_iterator<char>());
     bpf_skeleton ebpf_program;
-    if (ebpf_program.open_from_json_config(json_str, bpf_object_buffer) != 0) {
+    if (ebpf_program.open_from_json_config(json_str, bpf_object_buffer, nullptr) != 0) {
```
Comment (ocfox) [inline bpf-loader/src/main.cpp:29]:
```suggestion
    if (ebpf_program.open_from_json_config(json_str, bpf_object_buffer) != 0) {
```

File: bpf-loader/src/eunomia_bpf.cpp [hunk 6]
```diff
@@ -550,15 +553,18 @@ struct eunomia_bpf {
 };
 struct eunomia_bpf *
 open_eunomia_skel_from_json(const char *json_data,
-                            const char *bpf_object_buffer, size_t object_size)
+                            const char *bpf_object_buffer, size_t object_size,
+                            char *btf_archive_path)
```
Comment (ocfox) [inline bpf-loader/src/eunomia_bpf.cpp:557]:
```suggestion
                            const char *bpf_object_buffer, size_t object_size,
                            char *btf_archive_path = nullptr)
```

File: ecli/src/json_runner/json.rs [hunk 7]
```diff
@@ -44,6 +44,7 @@ pub fn handle_json(conf: ProgramConfigData) -> EcliResult<()> {
             json_data.as_ptr() as *const c_char,
             extra_arg_raw.as_mut_ptr(),
             extra_arg_raw.len() as i32,
+            std::ptr::null_mut(),
```
Comment (ocfox) [inline ecli/src/json_runner/json.rs:47]: Why optional argument still need a parameter?

**Review by yunwei37 — APPROVED** [review]

**Pull request merged**